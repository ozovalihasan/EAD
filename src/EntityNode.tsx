import { useCallback, useEffect } from 'react';
import { Handle, Position } from 'react-flow-renderer';
import TableName from './TableName';
import useStore from './zustand/store';


export type EntityNodeDataType = {
  tableId: string,
  name: string,
}

const EntityNode = ({id, data }: {id: string, data: EntityNodeDataType}) => {
  const { onNodeInputChange, onMouseEnterThrough, onChangeAssociationType } = useStore();

  const isConnectContinue = useStore(state => state.isConnectContinue)
  const isMouseOnNode = useStore(state => state.isMouseOnNode)
  const mouseOnNodeId = useStore(state => state.mouseOnNodeId)
  const selectedNodeForThrough = useStore(state => state.selectedNodeIdForThrough) == id
  const associationType = useStore(state => state.associationType)
  const connectionStartId = useStore(state => state.connectionStartId)

  const visibleSourceHandles = ((!isConnectContinue && isMouseOnNode && mouseOnNodeId == id) ? "visible" : "hidden")
  
  return (
    <div className={`border-black border border-solid p-1 rounded-sm ${ selectedNodeForThrough ?  "bg-second-400" : "bg-first-100"}`}>
      
      <Handle className="border-none w-6 h-6" type="target" position={Position.Top} id="top" style={{visibility: isConnectContinue ? "visible" : "hidden"}}/>
      
      <div>
        <label htmlFor="text"></label>
        <input value={data.name} className="bg-slate-200 w-32 p-1 rounded-md" id="text" name="text" onChange={(event) => onNodeInputChange(event, id)} />
      </div>
      <sub>
        {data && <TableName tableId={data.tableId}></TableName>}
      </sub>

      
      <Handle onMouseDown={() => onChangeAssociationType("has_one", id)} className="border-none w-6 h-6 left-1/4" style={{visibility: visibleSourceHandles}} type="source" position={Position.Bottom} id="bottom1" />
      <Handle onMouseDown={() => onChangeAssociationType("has_many", id)} className="border-none w-6 h-6" style={{visibility: visibleSourceHandles}} type="source" position={Position.Bottom} id="bottom2" />
      <Handle onMouseDown={() => onChangeAssociationType("through", id)} className="border-none w-6 h-6 left-3/4" style={{visibility: visibleSourceHandles}} type="source" position={Position.Bottom} id="bottom3" />

      {
        isConnectContinue && associationType === "through" && (connectionStartId !== id) && 
        <div className="text-tiny text-center text-first-100 bg-first-500  w-16 h-full absolute right-full bottom-0 rounded-l-md px-1 content-center flex items-center " onMouseEnter={(event) => onMouseEnterThrough(event, id)}>
          <div className='text-center w-full'>
            through
          </div>
        </div>
      }
    </div>
  );
}

export default EntityNode;

      