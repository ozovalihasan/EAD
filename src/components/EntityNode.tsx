import { useCallback, useEffect } from 'react';
import { Handle, Position } from 'react-flow-renderer';
import {
  HasManyHandle,
  HasOneHandle,
  TableName,
  ThroughHandle 
} from "."
import useStore from '../zustand/store';


export type EntityNodeDataType = {
  tableId: string,
  name: string,
}

export const EntityNode = ({id, data, selected }: {id: string, data: EntityNodeDataType, selected: boolean}) => {
  const { onNodeInputChange, onMouseEnterThrough, onChangeAssociationType } = useStore();

  const isConnectContinue = useStore(state => state.isConnectContinue)
  const isMouseOnNode = useStore(state => state.isMouseOnNode)
  const mouseOnNodeId = useStore(state => state.mouseOnNodeId)
  const selectedNodeForThrough = useStore(state => state.selectedNodeIdForThrough) == id
  const associationType = useStore(state => state.associationType)
  const connectionStartId = useStore(state => state.connectionStartId)

  const visibleSourceHandles = ((!isConnectContinue && isMouseOnNode && mouseOnNodeId == id) ? "visible" : "hidden")
  
  return (
    <div className={`border-black border border-solid p-1 rounded-sm ${ selectedNodeForThrough ?  "bg-second-400" : "bg-first-50"} ${ selected ?  "bg-first-200" : ""}`} >
      
      <Handle className="border-none w-6 h-6" type="target" position={Position.Top} id="top" style={{visibility: isConnectContinue ? "visible" : "hidden"}}/>
      
      <div>
        <label htmlFor="text"></label>
        <input placeholder='Entity' value={data.name} className="w-32 p-1 rounded-md" id="text" name="text" onChange={(event) => onNodeInputChange(event, id)} />
      </div>
        {data && <TableName tableId={data.tableId}></TableName>}

      
      
      <div className='relative w-full'>
        <HasOneHandle visibility={visibleSourceHandles}/>
        <Handle
          onMouseDown={() => onChangeAssociationType("has_one", id)}
          className="opacity-50 absolute -bottom-3 justify-center items-center flex border-none w-6 h-6 left-1/4"
          style={{visibility: visibleSourceHandles}}
          type="source"
          position={Position.Bottom}
          id="bottom1"
        />

        <HasManyHandle visibility={visibleSourceHandles}/>
        <Handle
          onMouseDown={() => onChangeAssociationType("has_many", id)}
          className=" opacity-50 absolute -bottom-3 justify-center items-center flex border-none w-6 h-6 left-1/2"
          style={{visibility: visibleSourceHandles}}
          type="source"
          position={Position.Bottom}
          id="bottom2"
        />

        <ThroughHandle visibility={visibleSourceHandles}/>
        <Handle
            onMouseDown={() => onChangeAssociationType("through", id)}
            className=" opacity-50 absolute -bottom-3 justify-center items-center flex border-none w-6 h-6 left-3/4"
            style={{visibility: visibleSourceHandles}}
            type="source"
            position={Position.Bottom}
            id="bottom3"
        />
      </div>

      {
        isConnectContinue && associationType === "through" && (connectionStartId !== id) && 
        <div className="text-tiny text-center text-first-100 bg-first-500  w-16 h-full absolute right-full bottom-0 rounded-l-md px-1 content-center flex items-center " onMouseEnter={(event) => onMouseEnterThrough(event, id)}>
          <div className='text-center w-full'>
            through
          </div>
        </div>
      }
    </div>
  );
}

      