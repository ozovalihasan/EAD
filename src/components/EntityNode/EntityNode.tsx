import { useCallback, useEffect } from 'react';
import { Handle, Position } from 'react-flow-renderer';
import {
  TableName,
  AllHandlers
} from "components"
import useStore from 'zustandStore/store';


export type EntityNodeDataType = {
  tableId: string,
  name: string,
}

export const EntityNode = ({id, data, selected }: {id: string, data: EntityNodeDataType, selected: boolean}) => {
  const {
    isConnectContinue,
    associationType,
    connectionStartId,
    selectedNodeIdForThrough,
    onNodeInputChange, 
    onMouseEnterThrough, 
  } = useStore();

  return (
    <div className={`border-black border border-solid p-1 rounded-sm ${ (selectedNodeIdForThrough == id) ?  "bg-second-400" : "bg-first-50"} ${ selected ?  "bg-first-200" : ""}`} >
      
      <Handle className="border-none w-6 h-6" type="target" position={Position.Top} id="top" style={{visibility: isConnectContinue ? "visible" : "hidden"}}/>
      
      <div>
        <label htmlFor="text"></label>
        <input placeholder='Entity' value={data.name} className="w-32 p-1 rounded-md" id="text" name="text" onChange={(event) => onNodeInputChange(event, id)} />
      </div>

      {data && <TableName tableId={data.tableId}></TableName>}
      
      <AllHandlers id={id} />

      {
        isConnectContinue && associationType === "through" && (connectionStartId !== id) && 
        <div className="text-tiny text-center text-first-100 bg-first-500  w-16 h-full absolute right-full bottom-0 rounded-l-md px-1 content-center flex items-center " onMouseEnter={(event) => onMouseEnterThrough(event, id)}>
          <div className='text-center w-full'>
            through
          </div>
        </div>
      }
    </div>
  );
}

      