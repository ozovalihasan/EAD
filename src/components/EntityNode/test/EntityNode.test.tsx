import { EntityNode } from '../EntityNode';
import { fireEvent, render, renderHook, screen } from "@testing-library/react";
import useStore from '@/zustandStore/store';


jest.mock('@/components',  () => ({
  TargetHandle: (
    ({ nodeId }: {nodeId: string} ) => (
      <>
        MockTargetHandle
        nodeId: {nodeId}
      </>
    )
  ),
  TableName: (
    ({ tableId }: {tableId: string} ) => (
      <>
        MockTableName
        tableId: {tableId}
      </>
    )
  ),
  AllHandlers: (
    ({ nodeId }: {nodeId: string} ) => (
      <>
        MockAllHandlers
        nodeId: {nodeId}
      </>
    )
  ),
  SelectThroughNode: (
    ({ nodeId }: {nodeId: string} ) => (
      <>
        MockSelectThroughNode
        nodeId: {nodeId}
      </>
    )
  ),
      
}))

let renderReadyComponent: JSX.Element;

beforeEach(() => {
  
  useStore.setState({ 
    onNodeInputChange: jest.fn() 
  })
  
  renderReadyComponent = (
    <EntityNode 
      id="111"
      data={{name: "mockName", tableId: "mockTableId"}}
      selected
    />
  );
});

describe('<EntityNode />', () => {
  
  it('renders the necessary components correctly', () => {
    render(renderReadyComponent );
    expect(screen.getByText(/MockTargetHandle/i)).toBeInTheDocument();
    expect(screen.getByText(/MockTableName/i)).toBeInTheDocument();
    expect(screen.getByText(/MockAllHandlers/i)).toBeInTheDocument();
    expect(screen.getByText(/MockSelectThroughNode/i)).toBeInTheDocument();
  });

  it('renders an input with a placeholder "Entity"', () => {
    render(renderReadyComponent );
    const { result } = renderHook(() => useStore());
    
    const inputElement = screen.getByPlaceholderText(/Entity/i)

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveValue("mockName");

    fireEvent.change(inputElement, {target: {value: 'mockInput'}});

    expect(result.current.onNodeInputChange).toHaveBeenCalledTimes(1);

  });
  

  it('renders correctly', () => {
    useStore.setState({ 
      associationType: 'has_many'
    })
    
    const renderedContainer = render(renderReadyComponent );
    expect(renderedContainer).toMatchSnapshot();
  });
  

});