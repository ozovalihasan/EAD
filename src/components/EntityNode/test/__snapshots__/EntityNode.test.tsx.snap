// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntityNode /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="
          p-1 rounded-md
          bg-first-500 dark:bg-first-600 
          bg-first-400 dark:bg-first-500
        "
      >
        MockTargetHandle nodeId: 
        111
        <input
          class="w-32 m-1 p-1 rounded-md ring-0 ring-offset-0"
          id="text"
          name="text"
          placeholder="Entity"
          tabindex="3"
          value="mockName"
        />
        MockTableName nodeId: 
        111
        tableId: 
        mockTableId
        MockAllHandlers nodeId: 
        111
        MockSelectThroughNode nodeId: 
        111
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="
          p-1 rounded-md
          bg-first-500 dark:bg-first-600 
          bg-first-400 dark:bg-first-500
        "
    >
      MockTargetHandle nodeId: 
      111
      <input
        class="w-32 m-1 p-1 rounded-md ring-0 ring-offset-0"
        id="text"
        name="text"
        placeholder="Entity"
        tabindex="3"
        value="mockName"
      />
      MockTableName nodeId: 
      111
      tableId: 
      mockTableId
      MockAllHandlers nodeId: 
      111
      MockSelectThroughNode nodeId: 
      111
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
